test/integration/common/StudioActions.kt

```diff
@@ -45,39 +45,72 @@ import java.nio.file.Path
 import kotlin.test.assertEquals
 import kotlin.test.assertTrue
-import kotlin.test.fail
 import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
 import mu.KotlinLogging
 
 object StudioActions {
     private val LOGGER = KotlinLogging.logger {}
 
-    suspend fun clickIcon(composeRule: ComposeContentTestRule, icon: Icon, delayMillis: Int = Delays.RECOMPOSE) {
-        clickText(composeRule, icon.unicode, delayMillis)
+    suspend fun clickIcon(composeRule: ComposeContentTestRule, icon: Icon) {
+        clickText(composeRule, icon.unicode)
     }
 
-    suspend fun clickText(composeRule: ComposeContentTestRule, text: String, delayMillis: Int = Delays.RECOMPOSE) {
+    suspend fun clickText(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilNodeWithTextIsClickable(composeRule, text)
         composeRule.onNodeWithText(text).performClick()
-        delayAndRecompose(composeRule, delayMillis)
+        delayAndRecompose(composeRule)
     }
 
-    fun clickAllInstancesOfText(composeRule: ComposeContentTestRule, text: String) {
+    suspend fun clickAllInstancesOfText(composeRule: ComposeContentTestRule, text: String) {
         val length = composeRule.onAllNodesWithText(text).fetchSemanticsNodes().size
         for (i in 0 until length) {
             composeRule.onAllNodesWithText(text)[i].performClick()
         }
+        delayAndRecompose(composeRule)
     }
 
-    fun clickAllInstancesOfIcon(composeRule: ComposeContentTestRule, icon: Icon) {
+    suspend fun clickAllInstancesOfIcon(composeRule: ComposeContentTestRule, icon: Icon) {
         clickAllInstancesOfText(composeRule, icon.unicode)
     }
 
-    fun assertNodeExistsWithText(composeRule: ComposeContentTestRule, text: String): SemanticsNodeInteraction {
-        return composeRule.onNodeWithText(text).assertExists()
+    suspend fun waitUntilNodeWithTextIsClickable(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilAssertionPasses(composeRule) {
+            composeRule.onNodeWithText(text).assertHasClickAction()
+        }
     }
 
-    fun assertNodeNotExistsWithText(composeRule: ComposeContentTestRule, text: String) {
-        return composeRule.onNodeWithText(text).assertDoesNotExist()
+    suspend fun waitUntilNodeWithTextExists(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilAssertionPasses(composeRule) {
+            composeRule.onNodeWithText(text).assertExists()
+        }
+    }
+
+    suspend fun waitUntilTrue(composeRule: ComposeContentTestRule, condition: () -> Boolean) {
+        composeRule.waitUntil(Delays.WAIT_TIMEOUT) {
+            runBlocking {
+                delayAndRecompose(composeRule)
+            }
+            try {
+                return@waitUntil condition()
+            } catch (e: Exception) {
+                return@waitUntil false
+            }
+        }
+    }
+
+    suspend fun waitUntilAssertionPasses(composeRule: ComposeContentTestRule, assertion: () -> Any) {
```

Comment:
Right now, we're losing the specifics of the assertion that failed, and how it failed. In the case of a non-trivial assert such as `assertEquals`, we're losing the actual value (that was not equal to expected)

Given that `waitUntil` is a suspending function that just throws an exception if it times out, we can wrap it in our own `try-catch` block (catching ONLY the specific exception type that it throws on a timeout), store the latest assertion error to a local variable, and throw that specific error - perhaps wrapped in our own kind of timeout exception, so we get "timeout of N ms exceeded waiting for assertion to pass. The last assertion failure was: {error}"

Reply:
```kt
    suspend fun waitUntilAssertionPasses(composeRule: ComposeContentTestRule, assertion: () -> Any) {
        var latestAssertionError: AssertionError? = null
        try {
            composeRule.waitUntil(Delays.WAIT_TIMEOUT) {
                runBlocking {
                    delayAndRecompose(composeRule)
                }
                try {
                    assertion()
                    return@waitUntil true
                } catch (e: AssertionError) {
                    latestAssertionError = e
                    return@waitUntil false
                }
            }
        } catch (e: ComposeTimeoutException) {
            throw AssertionError("Timeout of ${Delays.WAIT_TIMEOUT}ms exceeded waiting for assertion " +
                    "to pass. See the cause for the last assertion failure below.", latestAssertionError)
        }
    }
    ```